common_app/io.c

```diff
@@ -0,0 +1,185 @@
+    io_seproxyhal_display_default(element);
+}
+#endif  // HAVE_BAGL
+
+WEAK uint8_t io_event(uint8_t channel) {
+    (void) channel;
+
+    switch (G_io_seproxyhal_spi_buffer[0]) {
+        case SEPROXYHAL_TAG_BUTTON_PUSH_EVENT:
+#ifdef HAVE_BAGL
+            UX_BUTTON_PUSH_EVENT(G_io_seproxyhal_spi_buffer);
+#endif  // HAVE_BAGL
+            break;
+        case SEPROXYHAL_TAG_STATUS_EVENT:
+            if (G_io_apdu_media == IO_APDU_MEDIA_USB_HID &&  //
+                !(U4BE(G_io_seproxyhal_spi_buffer, 3) &      //
+                  SEPROXYHAL_TAG_STATUS_EVENT_FLAG_USB_POWERED)) {
+                THROW(EXCEPTION_IO_RESET);
+            }
+            /* fallthrough */
+        case SEPROXYHAL_TAG_DISPLAY_PROCESSED_EVENT:
+#ifdef HAVE_BAGL
+            UX_DISPLAYED_EVENT({});
+#endif  // HAVE_BAGL
+#ifdef HAVE_NBGL
+            UX_DEFAULT_EVENT();
+#endif  // HAVE_NBGL
+            break;
+#ifdef HAVE_NBGL
+        case SEPROXYHAL_TAG_FINGER_EVENT:
+            UX_FINGER_EVENT(G_io_seproxyhal_spi_buffer);
+            break;
+#endif  // HAVE_NBGL
+        case SEPROXYHAL_TAG_TICKER_EVENT:
+            UX_TICKER_EVENT(G_io_seproxyhal_spi_buffer, {});
+            break;
+        default:
+            UX_DEFAULT_EVENT();
+            break;
+    }
+
+    if (!io_seproxyhal_spi_is_status_sent()) {
+        io_seproxyhal_general_status();
+    }
+
+    return 1;
+}
+
+WEAK uint16_t io_exchange_al(uint8_t channel, uint16_t tx_len) {
+    switch (channel & ~(IO_FLAGS)) {
+        case CHANNEL_KEYBOARD:
+            break;
+        case CHANNEL_SPI:
+            if (tx_len) {
+                io_seproxyhal_spi_send(G_io_apdu_buffer, tx_len);
+
+                if (channel & IO_RESET_AFTER_REPLIED) {
+                    halt();
+                }
+
+                return 0;
+            } else {
+                return io_seproxyhal_spi_recv(G_io_apdu_buffer, sizeof(G_io_apdu_buffer), 0);
+            }
+        default:
+            THROW(INVALID_PARAMETER);
+    }
+
+    return 0;
+}
+
+/**
+ * Variable containing the length of the APDU response to send back.
+ */
+static uint32_t G_output_len = 0;
+
+/**
+ * IO state (READY, RECEIVING, WAITING).
+ */
+static io_state_e G_io_state = READY;
+
+WEAK void io_init() {
+    // Reset length of APDU response
+    G_output_len = 0;
+    G_io_state = READY;
+}
+
+WEAK int io_recv_command() {
+    int ret = -1;
+
+    switch (G_io_state) {
+        case READY:
+            G_io_state = RECEIVED;
+            ret = io_exchange(CHANNEL_APDU, G_output_len);
+            break;
+        case RECEIVED:
+            G_io_state = WAITING;
+            ret = io_exchange(CHANNEL_APDU | IO_ASYNCH_REPLY, G_output_len);
+            G_io_state = RECEIVED;
+            break;
+        case WAITING:
+            G_io_state = READY;
+            ret = -1;
+            break;
+    }
+
+    return ret;
+}
+
+WEAK int io_send_response(const buffer_t *rdata, uint16_t sw) {
```

Comment:
I'm not found of taking a `buffer_t` as parameter. Is it really better than a classical `uint8_t` pointer and associated buffer len?
Maybe we could at least expose both (`io_send_reponse_buffer()` and `io_send_response_pointer()`)?